Adds support for No Version Warning xml file

dave-gray101 opened the pull request:
While it's a minor issue, the "Version Mismatch" warning icons within RimSort were getting on my nerves for mods that are validated to work with 1.5 via [https://github.com/emipa606/NoVersionWarning](https://github.com/emipa606/NoVersionWarning)

This PR adds the ability to optionally use either a local copy or a url reference to https://raw.githubusercontent.com/emipa606/NoVersionWarning/refs/heads/main/1.5/ModIdsToFix.xml -- which contains a listing of mod ids that have been tested as compatible with the latest version of the game.

This is my first PR to RimSort - so please let me know if there are any style issues or other changes that need to be made.

I want to take a look at adding support for https://github.com/emipa606/UseThisInstead in a future PR.

dave-gray101 commented:
tagging issue #519 - this PR will address it.

LionelColaso commented:
Thank you for your interest and help just a few Suggestions
1) Since in Future you would want to add support for "Use This Instead" I recommend that you move this to its own Tab so that "Use This Instead" could also be added there
2) I see that you are directly downloading the raw file instead of cloning the link using Git, This means that you would have issue with uploading for which u implemented a notification
3) Please Replace the raw downloading of XML with git function of RimSort since it would be easy downloading and uploading, you can check how community database or steam database are cloned for reference

Patryk-Malinowski commented:
Something to consider is that MODS.md is present in the mod directory itself. You could maybe use it (as the local option), and if the No Version Warning mod gets updated, so does the MODS.md list.

dave-gray101 commented:
I did consider working off the downloaded from steam copy of the mod rather than their GitHub -- but honestly that seemed unnecessarily complex for users. I can revisit that if people would prefer it though!


Regarding the git versus single file -- I went that route due to the fact that I didn't see a need for any other files in that repository, but for consistency I have no argument against checking out the entire repository and digging out the file I need. Looking ahead to Use This Instead, I'll need to use that method anyway so I will make this adjustment soon.


Regarding the UI tab request: I'm happy to make that change now, but the tab will be pretty sparse looking until I complete Use This Instead. Does anyone have a preference if I migrate these settings to a new tab in this PR or in the follow-on Use This Instead?

LionelColaso commented:
> This is my first PR to RimSort - so please let me know if there are any style issues or other changes that need to be made.

If you missed anything check
https://rimsort.github.io/RimSort/development-guide/contributor-guidelines/#misc-coding-style

dave-gray101 commented:
> > This is my first PR to RimSort - so please let me know if there are any style issues or other changes that need to be made.
> 
> If you missed anything check https://rimsort.github.io/RimSort/development-guide/contributor-guidelines/#misc-coding-style

I've gone ahead and hit the altered file with Ruff.
I also added preliminary support for Use This Instead - it was relatively simple to integrate the "metadata / git" side of things at the same time that I was reworking the UI, but I am not 100% sure how to best surface a mod replacement to the user. For now, I've just added a simple message to the warning tooltip -but I'm open to other ideas as well.

LionelColaso commented:
> I've gone ahead and hit the altered files with Ruff. I also added preliminary support for Use This Instead - it was relatively simple to integrate the "metadata / git" side of things at the same time that I was reworking the UI, but I am not 100% sure how to best surface a mod replacement to the user. For now, I've just added a simple message to the warning tooltip -but I'm open to other ideas as well.

Just check #424 and #668
Its not the same but you will get the idea to some extent

dave-gray101 commented:
To expand a little on the linked issues 418 - 

I think what you're suggesting here is have a dynamic component in addition to KNOWN_MOD_REPLACEMENTS. This would allow the dependency checker to "replace" a dependency on the old mod with the new version? Based on what I know of RimWorld modding that seems safe to me, but I'd like to check!

If I'm misunderstanding, please let me know.

LionelColaso commented:
> To expand a little on the linked issues 418 -
> 
> I think what you're suggesting here is have a dynamic component in addition to KNOWN_MOD_REPLACEMENTS. This would allow the dependency checker to "replace" a dependency on the old mod with the new version? Based on what I know of RimWorld modding that seems safe to me, but I'd like to check!
> 
> If I'm misunderstanding, please let me know.

I think you could just add a New "Mod Replacement" button below "warnings/errors" and make it open a dialog showing the recommended mod replacement, something like #730

dave-gray101 commented:
> > To expand a little on the linked issues 418 -
> > I think what you're suggesting here is have a dynamic component in addition to KNOWN_MOD_REPLACEMENTS. This would allow the dependency checker to "replace" a dependency on the old mod with the new version? Based on what I know of RimWorld modding that seems safe to me, but I'd like to check!
> > If I'm misunderstanding, please let me know.
> 
> I think you could just add a New "Mod Replacement" button below "warnings/errors" and make it open a dialog showing the recommended mod replacement, something like #730

That's a good idea - although I'd be tempted to kick that to a followup PR - I think I need to really understand what's going on in `missing_mod_panel.py` before I duplicate how it works - the eval on line 219 gives me the creeps!

In particular, I also want to brainstorm what the buttons should actually... do. While it's easy to replicate the existing "download" functionality to grab the replacement mods, should we automatically unsubscribe from the old ones? Kick them out of the load order? Should those functions be automatic, or distinct buttons in the dialog? If we do unsubscribe from the mods, I'm frequently affected by #621 - so should that be contingent on that issue first?

LionelColaso commented:
> That's a good idea - although I'd be tempted to kick that to a followup PR - I think I need to really understand what's going on in `missing_mod_panel.py` before I duplicate how it works - the eval on line 219 gives me the creeps!

The "eval" on line "219" of "missing_mods_panel" is used for workshop mod updates it was done in #134 specifically in [Commit](https://github.com/RimSort/RimSort/pull/134/commits/03345a2fd719b47571a260658ab03401332ee0de#diff-27f0fd394abdfa2e73e52a359560fb22eccaed5d4baa3f1da3c67135bbad4ff8)

> In particular, I also want to brainstorm what the buttons should actually... do. While it's easy to replicate the existing "download" functionality to grab the replacement mods, should we automatically unsubscribe from the old ones? Kick them out of the load order? Should those functions be automatic, or distinct buttons in the dialog? If we do unsubscribe from the mods, I'm frequently affected by #621 - so should that be contingent on that issue first?

Unsubscribing the mod directly may cause issues, so better to avoid it

About #621 I never encountered such issues (i do not use steam) so could be related to .DDS files leftovers i am unsure

Give user "3" options when "Replace mod" button is clicked and use the existing missing mods panel buttons to get the mod the "2" option could be a bit tricky
1) Download the mod (just downloads the mod) 
2) Download and replace mod (moves new downloaded mod to "active list" while moving the old mod to "Inactive list")
3) Cancel, skip or something (does nothing and closes the window)

dave-gray101 commented:
I've made significant progress on the UI for this - although, based on a comment in the code I've done this by creating a common base class for MissingModPrompt, ModUpdaterPrompt, and my new UseThisInsteadPanel.

I know how to reliably trigger ModUpdaterPrompt in manual testing so that I can validate things there. Can anyone give me a simple manual procedure to trigger the Missing Mod Prompt workflow? I've never actually seen that screen in my "normal" RimSort operations. I mostly bring this up as things will be cleaner in this base class if I decide that all "BaseModsPanel" subclasses have the initial checkbox column rather than it being a subclass feature -- so I want to run some experiments to see what it would be like to have that panel also have checkboxes.

Patryk-Malinowski commented:
> I've made significant progress on the UI for this - although, based on a comment in the code I've done this by creating a common base class for MissingModPrompt, ModUpdaterPrompt, and my new UseThisInsteadPanel.
> 
> I know how to reliably trigger ModUpdaterPrompt in manual testing so that I can validate things there. Can anyone give me a simple manual procedure to trigger the Missing Mod Prompt workflow? I've never actually seen that screen in my "normal" RimSort operations. I mostly bring this up as things will be cleaner in this base class if I decide that all "BaseModsPanel" subclasses have the initial checkbox column rather than it being a subclass feature -- so I want to run some experiments to see what it would be like to have that panel also have checkboxes.

1. Have two isntances in RimSort. 
2. Change from one instance to the other.
3. You will get the missing mods prompt IF the second instance's active modlist is different

![image](https://github.com/user-attachments/assets/7551fc09-f1ca-4b8d-b29b-70b5d0180586)

Default Instance:
![image](https://github.com/user-attachments/assets/a14f92d0-d0b9-436f-8c52-221996297f92)

Dev Instance:
![image](https://github.com/user-attachments/assets/32652407-adcd-4cd3-a1c4-04c3644cdf5e)

dave-gray101 commented:
> > I've made significant progress on the UI for this - although, based on a comment in the code I've done this by creating a common base class for MissingModPrompt, ModUpdaterPrompt, and my new UseThisInsteadPanel.
> > I know how to reliably trigger ModUpdaterPrompt in manual testing so that I can validate things there. Can anyone give me a simple manual procedure to trigger the Missing Mod Prompt workflow? I've never actually seen that screen in my "normal" RimSort operations. I mostly bring this up as things will be cleaner in this base class if I decide that all "BaseModsPanel" subclasses have the initial checkbox column rather than it being a subclass feature -- so I want to run some experiments to see what it would be like to have that panel also have checkboxes.
> 
> 1. Have two isntances in RimSort.
> 2. Change from one instance to the other.
> 3. You will get the missing mods prompt IF the second instance's active modlist is different
> 
> ![image](https://private-user-images.githubusercontent.com/145374860/411784966-7551fc09-f1ca-4b8d-b29b-70b5d0180586.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMzIzNzMsIm5iZiI6MTczOTIzMjA3MywicGF0aCI6Ii8xNDUzNzQ4NjAvNDExNzg0OTY2LTc1NTFmYzA5LWYxY2EtNGI4ZC1iMjliLTcwYjVkMDE4MDU4Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMVQwMDAxMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02N2MzNGViODBkNjUwOGU1ZDc0OThkOWEwYzNhZjg5MGI0MzFkMTg1MzE5NGNhOTdjNGU4YzE3MjYzNzFhNDNlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pkuhiOIEx4T_cTa8bKjhwe6UOiCprXuxJnm2jXrP1c4)
> 
> Default Instance: ![image](https://private-user-images.githubusercontent.com/145374860/411785051-a14f92d0-d0b9-436f-8c52-221996297f92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMzIzNzMsIm5iZiI6MTczOTIzMjA3MywicGF0aCI6Ii8xNDUzNzQ4NjAvNDExNzg1MDUxLWExNGY5MmQwLWQwYjktNDM2Zi04YzUyLTIyMTk5NjI5N2Y5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMVQwMDAxMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMzNjNjZmYTYxNTE3MzQ0YmFmMjljYmQ2N2Y1ZGMxZTQ2ZmM2MTcyMzQxMzhjZjczNWQ1NDQ4YTI0NDA3MDUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ujbjqkzhWlI_4GX5GgFZ2k3g1LHy2-_IpRvA9R7Et1I)
> 
> Dev Instance: ![image](https://private-user-images.githubusercontent.com/145374860/411785004-32652407-adcd-4cd3-a1c4-04c3644cdf5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMzIzNzMsIm5iZiI6MTczOTIzMjA3MywicGF0aCI6Ii8xNDUzNzQ4NjAvNDExNzg1MDA0LTMyNjUyNDA3LWFkY2QtNGNkMy1hMWM0LTA0YzM2NDRjZGY1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMVQwMDAxMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZjViZjBlNWMwZGQ4Zjc1ZjVlMmNhMTg1ODcwMzJkMjBhZjQ0MTk1OTk2MDAzODU3NjEwNDkzOTE2MmUyMGRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.opE--cmWYbITZAmTSp0PNY6d4uOfdoia3LRHA0JL3Qk)

So, I'm new to the instances feature in general and I think I've set it up wrong somehow.
* Backed up my initial mod list via Save Mod List
* Backed up "Default" instance
* Created Dev instance, used autodetect to select the usual folders.
* Saw active mod list clear, got message about needing to run RimWorld once to get the mod list to populate correctly
* Ran RimWorld to do so, saved the active mod list
* Swapped to Default instance - the mod list **remained blank**
* Manually reloaded mod list from xml in Default instance
* Swapped to Dev instance - the mod list **remains the same as Default**


Swapping between them doesn't seem to really "work" the way I have configured things. Clearly this isn't right! Which folder should I change in the settings in my Dev instance in order to split things up properly ?

Patryk-Malinowski commented:
Use different paths for the dev instance. So for local mods just set it to some folder like dev_mods_test. Steam you can leave blank if I remember correctly.

dave-gray101 commented:
> Use different paths for the dev instance. So for local mods just set it to some folder like dev_mods_test. Steam you can leave blank if I remember correctly.
> 
> Then hopefully all you see is RimWorld and any dlc you have.

That seems to work! However, when I swap between these instances, I'm seeing this dialog:

![image](https://github.com/user-attachments/assets/8a6c4e7f-8b64-4d58-89f0-e299c9588240)


I was hoping to see a different one - [this one](https://github.com/RimSort/RimSort/blob/main/app/windows/missing_mods_panel.py)

Patryk-Malinowski commented:
> I was hoping to see a different one - [this one](https://github.com/RimSort/RimSort/blob/main/app/windows/missing_mods_panel.py)

Oh sorry. For that one I think you need this setting enabled:
![image](https://github.com/user-attachments/assets/cb3b1cdf-9934-441b-aa6b-fb7d738cb0d1)


In Advanced settings

dave-gray101 commented:
Thanks for your help! I'm able to get all the testing I need done here now.
Unless anyone strongly objects, I think I will add a checkbox to this Missing Mod screen - it's already hidden behind a checkbox on the advanced settings page, so I think users will tolerate some (small) functionality changes - and I think the tool itself will be more useful if it's possible to download files in chunks anyway.

dave-gray101 commented:
![image](https://github.com/user-attachments/assets/51e3189b-d416-4775-8c26-d4487ce8f1cf)

The UI isn't the greatest, but I think it covers the basics now. Currently, it only supports _downloading_ the replacement mods. Since in many cases the mod IDs will stay the same, I'm not sure if it's a good idea to automatically tamper with the mod list - in some of the cases I tested in this screenshot, the newer mod automatically gets activated.

I'm considering adding two more buttons to the UI to unsubscribe from the old versions directly from this page - if I add that, I'll post a new screenshot.


I'll update the PR description and title to be a bit more reflective of the larger scope of this PR soon.

Patryk-Malinowski commented:
Coming along really well! Great job so far :)

dave-gray101 commented:
![image](https://github.com/user-attachments/assets/88d1ae34-f931-44a8-ab59-90f2783b53da)

I think I'm happy with things now - let me know if there's any other suggestions.

LionelColaso commented:
> ![image](https://private-user-images.githubusercontent.com/15875595/412662577-88d1ae34-f931-44a8-ab59-90f2783b53da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0MTY3NjEsIm5iZiI6MTczOTQxNjQ2MSwicGF0aCI6Ii8xNTg3NTU5NS80MTI2NjI1NzctODhkMWFlMzQtZjkzMS00NGE4LWFiNTktOTBmMjc4M2I1M2RhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDAzMTQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4ODgyYzNkNjJkZjJmYzI2OGM4ZGViZjEyYWZkMzNhNjhkMzZmYzFhNzYxMjIxMTYyODczYjAyMmRlOTE4MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MdQopjr-rtqm9rzwJXCO9ElhYTrOI_j7dEDCI_dDDsc)
> 
> I think I'm happy with things now - let me know if there's any other suggestions.

If not asking too much would it be possible for you to update the wiki as well with your new changes? I can pe a separate pr

LionelColaso commented:
> ![image](https://private-user-images.githubusercontent.com/15875595/412662577-88d1ae34-f931-44a8-ab59-90f2783b53da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0MzE4OTIsIm5iZiI6MTczOTQzMTU5MiwicGF0aCI6Ii8xNTg3NTU5NS80MTI2NjI1NzctODhkMWFlMzQtZjkzMS00NGE4LWFiNTktOTBmMjc4M2I1M2RhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDA3MjYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5ZTZjNjlkYWFkNjcyNGM4OTZhZTRhNTUxYzg3OTBmYWVlYzhiYWNlMGJlMzU2MzVkNDAwMzNlOWYyOTAyZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZP71WdVOyQd1XCgkzSB-2rQ-IghGHRTTn4cjvIW4cpg)
> 
> I think I'm happy with things now - let me know if there's any other suggestions.

Just Rearrange the buttons
some things to consider
1) personally i dont think adding the Unsubscribe buttons, a good idea since it will result in old mod deletions,
consider a scenario in replacement mod has some issues and for some reason the game or save file does not load with the replacement mod
2) No doubt unsubscribe button is good idea but i would Let the users do that manually if they want to unsubscribe

the above is my personal opinion since i dont use steam so i dont know if its actually a good or bad idea
i would like to hear what others think as well

dave-gray101 commented:
I think I'd prefer to do the wiki changes in a separate PR - just to make things clearer for myself. I'll get started on that soon.

Regarding button rework:
I've had a new idea. Instead of having separate buttons for everything, I'm going to do a prototype that uses a "main" action button with a QToolButton next to it that opens a dropdown menu of other actions. This would let me move the unsubscribe buttons into that layer. I ended up adding those buttons for personal use - if one subscribes to the replacements and refreshes the interface, RimSort complains about duplicate mod IDs. Unsubscribing alone isn't sufficient to solve this problem however - it looks like actually deleting the mod is.

I was hesitant to take that step without "warning" - I like the idea of a confirmation box to solve that problem.

The "menu" design will let me free up the screen real estate to have distinct "unsubscribe" and "delete" options - the later of which I can pull from existing menus.

LionelColaso commented:
If it works for personal use and easy for normal users to understand and not be confusing, I would say go for it

Patryk-Malinowski commented:
Oh yeah the unsubscribe doesn't delete stuff currently. It needs to wait for steam file validation to be triggered manually/automatically. I am making some changes to the steam integration stuff to either delete the mods after unsubscribing or trigger file validation.

Patryk-Malinowski commented:
I definitely would keep the unsubscribe options there. I like the idea of actually **replacing** the outdated mods in as little clicks as possible. 

If the new replacement mod doesn't work we could deal with that as a user, but I feel like 9/10 times it will work. Also the `Open Workshop Page` button is right there for the user to check the replacement mod out and make sure it will work etc.

Having the Subscribe and Unsubscribe options separate is fair enough.

dave-gray101 commented:
![image](https://github.com/user-attachments/assets/ee468296-df67-402e-a94a-261cb76dde89)

![image](https://github.com/user-attachments/assets/48a93b07-4525-41c9-9685-3ea0119272e5)


I've updated this PR with a menu that holds unsub / deletion options.

LionelColaso reviewed (dismissed):
Looks good for merge
